File: framework/test2/list/revall/revall.rb
require 'set'
require 'soda/leetcode/list'
# step [1]: implement solution function
# @param {ListNode[]} lists
# @return {ListNode[]}
def reverse_all(lists)
  (0...lists.size).each { |i|
    lists[i] = reverse(lists[i])
  }
  i, j = 0, lists.size - 1
  while i < j
    lists[i], lists[j] = lists[j], lists[i]
    i += 1
    j -= 1
  end
  lists
end

def reverse(head)
  h = nil
  while head
    _next = head.next
    head.next = h
    h = head
    head = _next
  end
  h
end

if __FILE__ == $0
  require 'soda/unittest/work'
  ns = Soda::Unittest
  # step [2]: setup function/return/arguments
  work = ns::TestWork.create(method(:reverse_all))
  # work = ns::TestWork.forStruct(CLASS)
  # work.validator = -> (e, r) { e is equal to r }
  work.compare_serial = true
  puts work.run(ARGF.read)
end

File: framework/test2/list/revall/revall.pl
use strict;
use warnings;
use Soda::Leetcode::ListNode;

*ListNode:: = \*Soda::Leetcode::ListNode::;

# step [1]: implement solution function
# @param {ListNode[]} lists
# @return {ListNode[]}
sub reverse_all {
    my $lists = shift;
    for my $i (0..$#$lists) {
        $lists->[$i] = reverse_list($lists->[$i]);
    }
    my ($i, $j) = (0, $#$lists);
    while ($i < $j) {
        ($lists->[$i], $lists->[$j]) = ($lists->[$j], $lists->[$i]);
        ++$i;
        --$j;
    }
    $lists;
}

sub reverse_list {
    my $head = shift;
    my $h;
    while (defined $head) {
        my $next = $head->next;
        $head->set_next($h);
        $h = $head;
        $head = $next;
    }
    $h;
}

use Soda::Unittest::Utils qw(from_stdin);
use Soda::Unittest::TestWork qw(create);
use Soda::Unittest::Validators qw(for_array for_array2d);
# step [2]: setup function/return/arguments
my $work = create(\&reverse_all);
# $work->validator(sub { $_[0] vs $_[1]; return boolean });
$work->compare_serial(1);
print $work->run(from_stdin());

